### PR TITLE
Get `cargo clippy` clean

### PR DESCRIPTION
**Pull request by fitzgen**
Also `cargo fmt` with the latest `rustfmt`.

r? @philipc or @tromey

### CODE REVIEW

File: benches/bench.rs [hunk 1]
```diff
@@ -254,17 +260,18 @@ mod cfi {
                 match entry {
                     CieOrFde::Cie(cie) => {
                         let mut instrs = cie.instructions();
-                        while let Some(i) = instrs.next()
-                            .expect("Can parse next CFI instruction OK") {
+                        while let Some(i) =
+                            instrs.next().expect("Can parse next CFI instruction OK") {
```
Comment (jonhoo) [inline benches/bench.rs]: Ugh... But if rustfmt wants it...

File: benches/bench.rs [hunk 2]
```diff
@@ -318,10 +327,13 @@ mod cfi {
...
                           -> usize {
-        fde.instructions().fold(0, |count, _| count + 1).expect("fold over instructions OK")
+        fde.instructions()
+            .fold(0, |count, _| count + 1)
+            .expect("fold over instructions OK")
     }
 
-    fn get_fde_with_longest_cfi_instructions(eh_frame: EhFrame<LittleEndian>)
+    fn get_fde_with_longest_cfi_instructions
+        (eh_frame: EhFrame<LittleEndian>)
```
Comment (jonhoo) [inline benches/bench.rs]: Really? `rustfmt` wants that change?

Comment (fitzgen) [inline benches/bench.rs]: ¯\_(ツ)_/¯

File: rustfmt.toml [hunk 3]
```diff
@@ -1,2 +1,2 @@
 write_mode = "Overwrite"
-
+closure_block_indent_threshold = 1
```
Comment (jonhoo) [inline rustfmt.toml]: Ahh, this is why some of the changes are looking so odd to me. Why do you want this here?

File: src/abbrev.rs [hunk 4]
```diff
@@ -650,10 +652,8 @@ pub mod tests {
         let expect = Some(Abbreviation::new(1,
                                             constants::DW_TAG_subprogram,
                                             constants::DW_CHILDREN_no,
-                                            vec![
-                    AttributeSpecification::new(constants::DW_AT_name,
-                                                constants::DW_FORM_string),
-                ]));
+                                            vec![AttributeSpecification::new(constants::DW_AT_name,
+                                                                    constants::DW_FORM_string)]));
```
Comment (jonhoo) [inline src/abbrev.rs]: That's not great, but then again, rustfmt still doesn't deal with macros very well..

File: src/lib.rs [hunk 5]
```diff
@@ -148,6 +148,15 @@
...
 #![deny(missing_debug_implementations)]
 
+// Allow clippy warnings when we aren't building with clippy.
+#![allow(unknown_lints)]
+// False positives with `fallible_iterator`.
+#![allow(should_implement_trait)]
+// Many false positives involving `continue`.
+#![allow(never_loop)]
+// False positives when block expressions are used inside an assertion.
+#![allow(panic_params)]
```
Comment (jonhoo) [inline src/lib.rs]: These are a bit unfortunate. Are there upstream bugs filed against clippy for the worst offenders here that we can link to?

File: src/lookup.rs [hunk 6]
```diff
@@ -26,6 +26,7 @@ pub trait LookupParser<'input, Endian>
     /// Parse a header from `input`. Returns a tuple of `input` sliced beyond this header and
     /// all of its entries, `input` sliced to contain just the entries corresponding to this
     /// header (without the header itself), and the parsed representation of the header itself.
+    #[allow(type_complexity)]
```
Comment (jonhoo) [inline src/lookup.rs]: Clippy is right that a number of these are kind of unfortunate, but probably worth dealing with separately.

File: src/unit.rs [hunk 7]
```diff
@@ -975,6 +976,8 @@ impl<'input, Endian> Attribute<'input, Endian>
     /// name.
     ///
     /// See "Figure 20. Attribute encodings" and "Figure 21. Attribute form encodings".
+    #[allow(cyclomatic_complexity)]
+    #[allow(match_same_arms)]
```
Comment (jonhoo) [inline src/unit.rs]: Why is this necessary?

File: src/unit.rs [hunk 8]
```diff
@@ -4146,10 +4172,22 @@ mod tests {
...
+                    .expect("Should parse next sibling")
+                    .is_none());
+        assert!(cursor
+                    .next_sibling()
+                    .expect("Should parse next sibling")
+                    .is_none());
+        assert!(cursor
+                    .next_sibling()
+                    .expect("Should parse next sibling")
+                    .is_none());
```
Comment (jonhoo) [inline src/unit.rs]: :(

File: tests/parse_self.rs [hunk 9]
```diff
@@ -39,12 +39,14 @@ fn test_parse_self_debug_info() {
...
+        while cursor
+                  .next_dfs()
+                  .expect("Should parse next dfs")
+                  .is_some() {
             let entry = cursor.current().expect("Should have a current entry");
 
             let mut attrs = entry.attrs();
-            while let Some(_) = attrs.next().expect("Should parse entry's attribute") {
-            }
+            while let Some(_) = attrs.next().expect("Should parse entry's attribute") {}
```
Comment (jonhoo) [inline tests/parse_self.rs]: Shouldn't this be `.is_some()`?

File: src/cfi.rs [hunk 10]
```diff
@@ -629,22 +630,23 @@ impl<'bases, 'input, Endian, Section> FallibleIterator
...
-fn parse_cfi_entry_common<'input, Endian, Section>(input: EndianBuf<'input, Endian>)
-                                                   -> Result<Option<(EndianBuf<'input, Endian>,
-                                                                     (u64,
-                                                                      Format,
-                                                                      EndianBuf<'input, Endian>,
-                                                                      u64,
-                                                                      EndianBuf<'input, Endian>))>>
+/// `Ok(Some(tuple))`, where `tuple.0` is the start of the next entry and
+/// `tuple.1` is the parsed CFI entry data.
+fn parse_cfi_entry_common<'input, Endian, Section>
```
Comment (jonhoo) [inline src/cfi.rs]: Curious that we only call this from one place (at least as far as I can tell from what places had to be modified to use `CfiEntryCommon`). I guess it'll be inlined though, so it's okay.

**Comment by jonhoo** [comment]
Left some inline comments, but generally looks good to me. Biggest questions are:

 - Why the `rustfmt.toml` change?
 - Can we link to the worst clippy bugs so we can remove the `allow()` when they're fixed?

File: rustfmt.toml (hunk 3 above)
Comment (philipc) [inline rustfmt.toml]: I personally prefer block indent, so I'm fine with this change, and I also think the official rust style is moving towards preferring block indent too.

However, I expect the reason is that without this, rustfmt seems to be buggy and formats some stuff horribly, particularly in unit.rs. eg
````
             constants::DW_FORM_addr => {
                 return parse_address(input, unit.address_size()).map(|(rest, addr)| {
-                    let attr = Attribute {
+                                                                         let attr = Attribute {
                         name: spec.name(),
                         value: AttributeValue::Addr(addr),
                     };
-                    (rest, attr)
-                });
+                                                                         (rest, attr)
+                                                                     });
             }
````

File: src/abbrev.rs (hunk 4 above)
Comment (philipc) [inline src/abbrev.rs]: Generally rustfmt ignores macros if it can't understand them. I thought it had special handling for `vec!` though. It's an improvement over what was there before at least.

File: src/unit.rs (hunk 8 above)
Comment (philipc) [inline src/unit.rs]: Setting `chain_one_line_max = 100` would avoid this. I also like `fn_call_width = 100`.

File: rustfmt.toml (hunk 3 above)
Comment (jonhoo) [inline rustfmt.toml]: I also prefer block indent, but it does mean that we end up with block indents in a couple of places where rustfmt could choose not to wrap and indent at all. The issue you're referring to I believe was fixed by https://github.com/rust-lang-nursery/rustfmt/pull/1495, at least since then I haven't been seeing those format errors.

**Comment by fitzgen** [comment]
I'm going to drop the `rustfmt` changes for now, since they seem mildly controversial, whereas teh clippy ones don't seem to be.

File: tests/parse_self.rs (hunk 9 above)
Comment (fitzgen) [inline tests/parse_self.rs]: I think incremental compilation + clippy results in disappearing warnings.

File: src/unit.rs (hunk 7 above)
Comment (fitzgen) [inline src/unit.rs]: Some of the match clauses have the same arms, so clippy wants them to be a single match clause joined by `|`. That isn't appropriate in this case, IMO, although usually it is.

File: rustfmt.toml (hunk 3 above)
Comment (fitzgen) [inline rustfmt.toml]: The rustfmt.toml changes were there to prevent even more of this, although I'm not sure why this particular case still occurred.

Comment (fitzgen) [inline rustfmt.toml]: Er wait. The diff @philipc posted is if you *don't* have this line, and is an example of why I did add it. Sorry misread the comments.

File: src/lib.rs (hunk 5 above)
Comment (fitzgen) [inline src/lib.rs]: Will look into filing upstream bugs / seeing if these are known problems.

**Comment by philipc** [comment]
> I'm going to drop the rustfmt changes for now, since they seem mildly controversial, whereas teh clippy ones don't seem to be.

I would like to use rustfmt regardless of whether its output has some quirks (we've lived with them in the past).  However, I tried rustfmt 0.8.4, and it introduces compilation errors now...

**Comment by fitzgen** [comment]
> I would like to use rustfmt regardless of whether its output has some quirks (we've lived with them in the past).

I agree.

>  However, I tried rustfmt 0.8.4, and it introduces compilation errors now...

Yikes! I can repro, filed an issue: https://github.com/rust-lang-nursery/rustfmt/issues/1533
